rust/otap-dataflow/crates/otap/src/otap_batch_processor.rs

```diff
@@ -0,0 +1,922 @@
+    async fn process(
+        &mut self,
+        msg: Message<OtapPdata>,
+        effect: &mut local::EffectHandler<OtapPdata>,
+    ) -> Result<(), EngineError> {
+        match msg {
+            Message::Control(ctrl) => {
+                match ctrl {
+                    otap_df_engine::control::NodeControlMsg::TimerTick { .. } => {
+                        // Follow OTLP semantics: only flush on timer if a size threshold has been reached
+                        if self.size_triggers_emission() {
+                            self.flush_current(effect).await
+                        } else {
+                            Ok(())
+                        }
+                    }
+                    otap_df_engine::control::NodeControlMsg::Config { .. } => Ok(()),
+                    otap_df_engine::control::NodeControlMsg::Shutdown { .. } => {
+                        // Flush and shutdown
+                        self.flush_current(effect).await?;
+                        effect.info(LOG_MSG_SHUTTING_DOWN).await;
+                        Ok(())
+                    }
+                    otap_df_engine::control::NodeControlMsg::Ack { .. }
+                    | otap_df_engine::control::NodeControlMsg::Nack { .. }
+                    | otap_df_engine::control::NodeControlMsg::CollectTelemetry { .. } => Ok(()),
+                }
+            }
+            Message::PData(data) => {
+                let max = self.config.send_batch_max_size;
+                match OtapArrowRecords::try_from(data.clone()) {
+                    Ok(rec) => {
+                        let rows = rec.batch_length();
+                        // Per-signal policy: drop zero-row; pre-flush if exceeding max; append rows; flush on max or send_batch_size.
+                        match &rec {
+                            OtapArrowRecords::Logs(_) => {
+                                if rows == 0 {
+                                    effect.info(LOG_MSG_DROP_NON_OTAP).await;
+                                    Ok(())
+                                } else {
+                                    if max > FOLLOW_SEND_BATCH_SIZE_SENTINEL
+                                        && self.rows_logs + rows > max
+                                    {
+                                        self.flush_logs(effect).await?;
+                                    }
+                                    self.rows_logs += rows;
+                                    self.current_logs.push(rec);
+                                    if (max > FOLLOW_SEND_BATCH_SIZE_SENTINEL
+                                        && self.rows_logs >= max)
+                                        || self.rows_logs >= self.config.send_batch_size
+                                    {
+                                        self.flush_logs(effect).await
+                                    } else {
+                                        Ok(())
+                                    }
```

Comment:
This is kind of a continuation of this comment related to the semantics of `send_batch_size`. https://github.com/open-telemetry/otel-arrow/pull/1015/files#r2310873042

I don't think we need the second check `max > FOLLOW_SEND_BATCH_SIZE_SENTINEL && self.rows_logs >= max` because we've already done a check just above `max > FOLLOW_SEND_BATCH_SIZE_SENTINEL && self.rows_logs + rows > max` then added `rows` to `self.rows_logs`.

Also, if the user has explicitly set the batch size to zero, I don't think we want to flush here. So if/when config.send_batch_size type is changed to `Option<usize>`, we can probably simplify the second `if` block to:
```
if matches!(send_batch_size, Some(s) if rows > s) {
    self.flush_logs(effect).await
} else {
    Ok(())
}
```

Reply:
ah OK, I think I understand now. Thanks for clarifying